### ISSUE TITLE
MLlib API

### ISSUE THREAD

**Issue by senor-hadoop** [yieldbot/flambo#59]
Flambo lacks an API for MLlib.

I recall bringing this up before (I think) and was told that the API exists, but was not released. I would rather avoid reinventing the wheel, so if it does exist would very much appreciate its addition to Flambo.

**Comment by UltimateHurl** [comment]
I'd also love to see this. The Java inter-op is mostly painless, but I'd like to keep things looking Clojurific where possible :)

**Comment by mikeivanov** [comment]
I've started shoveling my stuff into a branch here, if anybody is interested - https://github.com/yieldbot/flambo/compare/develop...mikeivanov:mllib

Comments, suggestions?

**Comment by senor-hadoop** [comment]
We should coordinate efforts then. I've also started on an ml and mllib implementation. I suggest we give Yieldbot a week or so more to respond; if nothing then we can open up an issue on one of our forks to discuss the additions.

**Comment by senor-hadoop** [comment]
@mikeivanov Good start. Did you look at any of the ML/Dataframes API? I've got an immediate need for a linear regression model. I'm beginning to think the ML package is too immature at this stage to be useful. Could you perhaps add an example of this using your API? Some unit tests too?

You might also be interested in the SQL changes I've made. There's a pull request waiting to be merged.

**Comment by mikeivanov** [comment]
@senor-hadoop yes, that was my impression, too - ML is too immature to rely upon. Basically, I'm working mostly with GLMs at work and that determined what I addressed first. I'm going to add some examples and tests.

**Comment by sorenmacbeth** [comment]
Howdy folks,

I've thought about this a bit, and I think what I'd like to do keep MLLib stuff out of flambo proper and make it a separate library. I have a couple other small libraries like this, one for parquet read/write support for example.

There are two way we could go

1) a completely separate repo(s) with pointers to other repos in the readme
2) an official "flambo-contrib" container repository which could hold all of the contrib libraries.

I've done the second option in past and it worked well enough.

If anyone wants to try and make a compelling argument why it _should_ be part of flambo proper, I'd be happy to listen and consider that as well.

No matter where we put it, I'm thrilled to have continued contributions to flambo, so thanks!

**Comment by mikeivanov** [comment]
+1 for flambo-contrib. Rationale: if later on you decide to merge it into flambo proper, it would be much easier to do that compared to pointers.

**Comment by senor-hadoop** [comment]
flambo-contrib works for me. @sorenmacbeth, does Yieldbot have anything from mllib to contribute, or are we starting from scratch? In our conversation back in August 2014 I think you mentioned there were existing mllib wrappers. I'd like to get to work on the linear regression functions if they don't exist already.

**Comment by sorenmacbeth** [comment]
@senor-hadoop we originally ported over some of the stuff in MLLib to clojure, but found the implementations questionable, so we created our own internal library from scratch, build on top of flambo. The plan is still to open source that, once it's more generic and has better tests and things. So, you're probably better off starting from scratch I think.

I can go ahead and create a `yieldbot/flambo-contrib` public repo soon.

**Comment by senor-hadoop** [comment]
You know, there is a reason to include mllib in flambo proper: it is in Spark proper. Thinking about how this would be used, it seems odd to place one of the Spark components in a contrib repository and all of the others in flambo.

What's the thinking behind just mllib being kept in a contrib repository?

**Comment by chetmancini** [comment]
Mostly agreeing here. I get the argument that MLLib is part of Spark just like Streaming or SQL, so maybe some basic functions belong here. I would note that every real machine learning pipeline feels very different. To make something useful to everyone I'm not sure how much value would actually be added. It's also really not that hard to do basic machine learning in flambo right now. I've found the complexity in my code comes from managing algorithms and configurations, massaging data, getting the output formats that I want, etc...ends up being a lot of RDD work.

Also agreeing MLLib to be fairly immature in some ways. For example, the new pipelines feature is pretty inadequate for much beyond the wordcount type documentation examples.

I have some protocols for configuring, selecting, and running different MLLib algos that feels pretty Clojure-y. Maybe something like that could be useful in `flambo-contrib`.

**Comment by senor-hadoop** [comment]
The work I am doing at the moment can probably enlighten this discussion somewhat, and also lead onto the broader topic of some refactoring that is going to be required in the near future if flambo is going to grow along with Spark. For example, using the API as it is, this is how we compute training error:

(def mse (-> (f/map data (f/fn [p] [(.label p) (.predict lr-model (.features p))])) ; Vector of [label,prediction]
             (f/map (f/fn [[l p]] (Math/pow (- l p) 2.0))) ; Square of the differences
             (.rdd)                     ; JavaRDD -> RDD
             (JavaDoubleRDD/fromRDD)
             (.mean))

And this is what it should be:

(def mse (-> (f/map data (f/fn [p] [(.label p) (.predict lr-model (.features p))]))
             (f/map-to-double (f/fn [[l p]] (Math/pow (- l p) 2.0)))
             (mean))

Those middle two lines are to work around the lack of map-to-double<https://spark.apache.org/docs/latest/api/java/org/apache/spark/api/java/JavaRDDLike.html#mapToDouble(org.apache.spark.api.java.function.DoubleFunction)>. It is easy enough to add, but requires changes to functions.clj and api.clj, and that’s where the potential refactoring may come in, since api.clj is going to quickly become unwieldy as we start implementing more Spark functions. This isn’t the main reason for bring it up here though. What is important is that map-to-double, mean, et. al. are all things that should be in flambo proper. The same goes for some of the functions @mikeivanov has added wrapping mllib. Anyone using Mllib will need  dense-vector and labeled-point, and these should go into flambo, not a contrib repo.

Where it gets a bit grey in my mind is things like read-csv. I added this sql.clj, where data frames lives because it seems to make the most sense to me (no doubt influenced by my R experience). I personally don’t see flambo as a strict wrapper but more as a computing DSL that happens to use Spark, just like I also use PigPen<https://github.com/Netflix/PigPen> as a Clojure wrapper for Pig. PigPen doesn’t just wrap Pig, it is its own DSL for map-reduce that uses Pig for the implementation. In other words, I think it’s OK to diverge from just wrapping Spark and build something that’s independently useful.

So, we could look at this two ways:

  *   Anything not a pure wrapper should go in contrib (like read-csv)
  *   Things that ‘make sense’, should go where it makes sense and contrib should be for useful add-ons that are somewhat complete and modular

Either way though, we should avoid the bike shed<http://en.wikipedia.org/wiki/Wikipedia:Avoid_Parkinson%27s_Bicycle_Shed_Effect> effect; I’m happy either way. The important thing is to keep the contributions coming.

Another thing we should do is get a ‘statement of direction’, or some such indication from yieldbot on their role going forward. Whilst I appreciate their making it public so we can build upon flambo, it is not clear how much effort they will be putting in going forward. Some strong (or any) leadership will definitely help us out. We’ve got three people recently identified that are working with flambo and willing to contribute, but we’re waiting on pull requests and ‘decisions’ on things like contrib vs. main, etc. Ideally yieldbot would commit to some number of hours per week, or at least a monthly pull-and-review so we can keep things moving. I know yieldbot are busy, but with a few more contributors, flambo becomes more useful for everyone. It would be time well spent. Barring that, perhaps adding a few committers to the repository would keep things moving. Worst case, we could all start working in one of our own repos if things remain slow in this repository. We should avoid the lisp curse<http://www.winestockwebdesign.com/Essays/Lisp_Curse.html>, where we all go off and independently develop what we need.

Any thoughts or comments?



From: Chet
Reply-To: yieldbot/flambo
Date: Thursday, May 7, 2015 at 08:36
To: yieldbot/flambo
Cc: Steve Nunez
Subject: Re: [flambo] MLlib API (#59)


Mostly agreeing here. I get the argument that MLLib is part of Spark just like Streaming or SQL, so maybe some basic functions belong here. I would note that every real machine learning pipeline feels very different. To make something useful to everyone I'm not sure how much value would actually be added. It's also really not that hard to do basic machine learning in flambo right now. I've found the complexity in my code comes from managing algorithms and configurations, massaging data, getting the output formats that I want, etc...ends up being a lot of RDD work.

Also agreeing MLLib to be fairly immature in some ways. For example, the new pipelines feature is pretty inadequate for much beyond the wordcount type documentation examples.

I have some protocols for configuring, selecting, and running different MLLib algos that feels pretty Clojure-y. Maybe something like that could be useful in flambo-contrib.
